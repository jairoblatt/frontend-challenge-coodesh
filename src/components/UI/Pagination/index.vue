<template>
  <div class="pagination" :class="{ 'pagination--active': this.loading }">
    <transition name="slide-up">
      <Spinner v-show="loading" size="40" />
    </transition>
  </div>
</template>

<script lang="ts">
import Vue, { PropOptions, PropType } from "vue";
import { sleep } from "@/utils/index";

export default Vue.extend({
  props: {
    distance: {
      type: Number,
      default: 40,
    } as PropOptions<number>,

    delay: {
      type: Number,
      default: 1000,
    } as PropOptions<number>,

    loading: Boolean as PropType<boolean>,
    forceInit: Boolean as PropType<boolean>,
  },

  data: () => ({
    ready: true,
  }),

  mounted() {
    this.forceInit && this.onLoadMore();
    document.addEventListener("scroll", this.infiniteScroll);
  },

  beforeDestroy() {
    document.removeEventListener("scroll", this.infiniteScroll);
  },

  methods: {
    async infiniteScroll() {
      const scroll = document.documentElement.clientHeight + document.documentElement.scrollTop;

      if (scroll >= document.documentElement.scrollHeight - this.distance && this.ready) {
        this.ready = false;
        this.onLoadMore();

        await sleep(this.delay);
        this.ready = true;
      }
    },

    onLoadMore() {
      this.$emit("loadMore");
    },
  },
});
</script>

<style lang="postcss" scoped>
.pagination {
  @apply w-full fixed bottom-0 left-2/4 transform -translate-x-2/4;
}

.pagination--active {
  @apply transition-colors;

  height: 10rem;
  background: linear-gradient(0deg, rgba(199, 199, 199, 0.548) 0%, rgba(255, 255, 255, 0) 86%);
}

.slide-enter,
.slide-leave-to {
  opacity: 0;
  transform: translateY(100%);
}

.slide-enter-active,
.slide-leave-active {
  transition: opacity 0.2s, transform 0.2s ease-in-out;
}
</style>
