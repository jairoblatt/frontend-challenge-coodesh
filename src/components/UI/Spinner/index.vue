<template>
  <div class="spinner-container">
    <span class="spinner" :style="styles"></span>
  </div>
</template>
<script lang="ts">
import Vue, { PropOptions } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Spinner",

  props: {
    size: {
      type: [String, Number],
      default: "inherit",
    } as PropOptions<string | number>,
  },

  computed: {
    styles(): CSS.Properties {
      return {
        minHeight: `${this.size}px`,
        minWidth: `${this.size}px`,
      };
    },
  },
});
</script>
<style lang="postcss" scoped>
.spinner-container {
  @apply relative flex justify-center items-center min-h-full min-w-full;
}

.spinner {
  @apply pointer-events-none border-solid rounded-full  border-primary-base  p-1;

  border-width: 0.3rem;
  border-top-color: white;
  animation: spinner 0.5s linear infinite;
}

@keyframes spinner {
  100% {
    transform: rotate(360deg);
  }
}
</style>
