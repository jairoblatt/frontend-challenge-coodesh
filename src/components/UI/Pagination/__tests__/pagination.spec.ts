import { mount, MountOptions, Wrapper, createLocalVue } from "@vue/test-utils";
import Pagination from "../index.vue";
import Spinner from "@/components/UI/Spinner/index.vue";

describe("Pagination unit tests", () => {
  type Instance = InstanceType<typeof Pagination>;
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>;

  const localVue = createLocalVue();

  localVue.component("Spinner", Spinner);

  beforeEach(() => {
    mountFunction = (config = {}) => {
      return mount(Pagination, {
        ...config,
        localVue,
      });
    };
  });

  it("Should mount the component", () => {
    const wrapper = mountFunction();

    expect(wrapper.vm).toBeDefined();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Should apply the "pagination--active" css class when the loading prop is true', () => {
    const wrapper = mountFunction({
      propsData: {
        loading: true,
      },
    });

    const wrapperElement = wrapper.findAll("div").at(0);

    expect(wrapperElement.classes("pagination--active")).toBe(true);
  });

  it("Should render spinner component when the loading prop it is true", () => {
    const wrapper = mountFunction({
      propsData: {
        loading: true,
      },
    });

    const spinnerComponent = wrapper.findComponent(Spinner);
    expect(spinnerComponent.attributes("style")).not.toBe("display: none;");
  });

  it('Should emit "loadMore" when the function it is called', () => {
    const wrapper = mountFunction();

    (wrapper.vm as any).onLoadMore();

    expect(wrapper.emitted("loadMore")).toBeTruthy();
    expect(wrapper.emitted("loadMore")?.length).toBe(1);
  });
});
