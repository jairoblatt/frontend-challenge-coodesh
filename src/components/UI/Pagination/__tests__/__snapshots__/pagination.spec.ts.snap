// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination unit tests Should mount the component 1`] = `
"<div class=\\"pagination\\">
  <transition-stub name=\\"slide-up\\">
    <div class=\\"spinner-container\\" style=\\"display: none;\\"><span class=\\"spinner\\" style=\\"min-height: 40px; min-width: 40px;\\"></span></div>
  </transition-stub>
</div>"
`;
